### ISSUE TITLE
Huge memory blowout in 0.11.1-RC1 with LZ4 compression

### ISSUE THREAD

**Issue by tudor** [edenhill/librdkafka#1445]
Description
===========

We see a huge memory blowout when using librdkafka 0.11.1-RC1. (C++ consumers using librdkafka, Java producers, broker 0.11) (hundreds of times more memory than we expect)

The cause is apparent in the code, in `rd_kafka_lz4_decompress`:

```
        /* If uncompressed size is unknown or out of bounds make up a
         * worst-case uncompressed size
         * More info on max size: http://stackoverflow.com/a/25751871/1821055 */
        if (fi.contentSize == 0 || fi.contentSize > inlen * 255)
                estimated_uncompressed_size = inlen * 255;
        else
                estimated_uncompressed_size = (size_t)fi.contentSize;
```

`fi.contentSize` is zero, because the Java code uses a streaming API (that doesn't set the uncompressed content size -- it doesn't know it, as it's streaming), so librdkafka has to be conservative and assume a worst-case compression ration of 255.

But, with multi-megabyte MessageSets, that means that librdkafka is allocating GBs of memory that are wasted; the real-life compression ratio in our case is more like 2, not 255. :)

The proper solution would probably be to allocate `inlen * 2` and grow as needed, but the `realloc` code further down is also broken:

```
                /* Need to grow output buffer, this shouldn't happen if
                 * contentSize was properly set. */
                if (unlikely(r > 0 && out_of == outlen)) {
                        char *tmp;
                        size_t extra = (r > 1024 ? r : 1024) * 2;

                        rd_atomic64_add(&rkb->rkb_c.zbuf_grow, 1);

                        if ((tmp = rd_realloc(outbuf, outlen + extra))) {
                                rd_rkb_log(rkb, LOG_WARNING, "LZ4DEC",
                                           "Unable to grow decompression "
                                           "buffer to %zd+%zd bytes: %s",
                                           outlen, extra,rd_strerror(errno));
                                err = RD_KAFKA_RESP_ERR__CRIT_SYS_RESOURCE;
                                goto done;
                        }
                        outlen += extra;
                }
```

That needs to realloc `out`, not `outbuf`, and it needs to fail if `tmp` is `NULL` (it fails if `tmp` is non-`NULL` instead), something like:

```
                        if (!(tmp = rd_realloc(out, outlen + extra))) {
                                rd_rkb_log(rkb, LOG_WARNING, "LZ4DEC",
                                           "Unable to grow decompression "
                                           "buffer to %zd+%zd bytes: %s",
                                           outlen, extra,rd_strerror(errno));
                                err = RD_KAFKA_RESP_ERR__CRIT_SYS_RESOURCE;
                                goto done;
                        } else {
                                out = tmp;
                        }
```

How to reproduce
================

Checklist
=========
Please provide the following information:

 - [x] librdkafka version (release number or git tag): 0.11.1-RC1
 - [x] Apache Kafka version: 0.11.0
 - [ ] librdkafka client configuration:
 - [x] Operating system: Ubuntu 16.04
 - [x] Using the legacy Consumer
 - [ ] Using the high-level KafkaConsumer
 - [ ] Provide logs (with `debug=..` as necessary) from librdkafka
 - [ ] Provide broker log excerpts
 - [ ] Critical issue

**Comment by tudor** [comment]
cc: @igorcanadi

**Comment by tudor** [comment]
I see that the fix for this (#1446) is not in 0.11.1. Are you planning to backport it into the 0.11 branch? I'd like to be able to run a released version in production if possible :) Thanks!

**Comment by edenhill** [comment]
This fix did not make 0.11.1 due to time constraints unfortunately, but we'll have a new release in about a month that will include this fix.
